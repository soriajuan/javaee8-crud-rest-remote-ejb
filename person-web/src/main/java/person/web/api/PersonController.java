package person.web.api;

import java.net.URI;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

import javax.ejb.EJB;
import javax.enterprise.context.RequestScoped;
import javax.inject.Inject;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import person.common.pojo.Person;
import person.common.service.PersonService;

@Path("person")
@RequestScoped
public class PersonController {

	@EJB(lookup = "java:global/person-backend-ear/person-ejb/PersonBean!person.common.service.PersonService")
	private PersonService personService;

	@Inject
	private PersonMapper mapper;

	@PUT
	@Path("/{id}")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response put(@PathParam("id") int id, PersonPatchRequestPayload payload) {

		// Not very elegant but better than @NotNull leaking data in response body or
		// filter with a cumbersome implementation
		if (Objects.isNull(payload)) {
			return Response.status(Response.Status.BAD_REQUEST).build();
		}

		Person updated = personService.update(id, mapper.toPerson(payload));
		PersonGetResponsePayload responsePayload = mapper.toPersonGetResponsePayload(updated);

		return Response.ok().entity(responsePayload).build();
	}

	@GET
	@Path("/{id}")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getById(@PathParam("id") int id) {
		return Response.ok().entity(mapper.toPersonGetResponsePayload(personService.findById(id))).build();
	}

	@DELETE
	@Path("/{id}")
	public void deleteById(@PathParam("id") int id) {
		personService.deleteById(id);
	}

	@GET
	@Produces(MediaType.APPLICATION_JSON)
	public Response getAll() {
		List<PersonGetResponsePayload> all = personService.findAll().stream().map(d -> mapper.toPersonGetResponsePayload(d))
				.collect(Collectors.toList());
		return Response.ok().entity(all).build();
	}

	@POST
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response post(PersonPostRequestPayload payload) {

		// Not very elegant but better than @NotNull leaking data in response body or
		// filter with a cumbersome implementation
		if (Objects.isNull(payload)) {
			return Response.status(Response.Status.BAD_REQUEST).build();
		}

		Person inserted = personService.insert(mapper.toPerson(payload));
		PersonGetResponsePayload responsePayload = mapper.toPersonGetResponsePayload(inserted);

		return Response.created(URI.create(String.format("/%d", responsePayload.getId()))).entity(responsePayload)
				.build();
	}

}
