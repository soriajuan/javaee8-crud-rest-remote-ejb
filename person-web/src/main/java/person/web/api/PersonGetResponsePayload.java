package person.web.api;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class PersonGetResponsePayload {

	private Integer id;
	private String firstName;
	private String lastName;

}
