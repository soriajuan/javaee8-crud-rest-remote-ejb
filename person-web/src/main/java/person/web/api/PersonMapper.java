package person.web.api;

import org.mapstruct.Mapper;
import org.mapstruct.Mapping;

import person.common.pojo.Person;

@Mapper(componentModel = "cdi")
interface PersonMapper {

	@Mapping(target = "id", ignore = true)
	Person toPerson(PersonPostRequestPayload p);
	
	@Mapping(target = "id", ignore = true)
	Person toPerson(PersonPatchRequestPayload p);

	PersonGetResponsePayload toPersonGetResponsePayload(Person p);

}
