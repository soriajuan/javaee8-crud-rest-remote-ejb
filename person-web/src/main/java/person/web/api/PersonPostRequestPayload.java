package person.web.api;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class PersonPostRequestPayload {

	private String firstName;
	private String lastName;

}
