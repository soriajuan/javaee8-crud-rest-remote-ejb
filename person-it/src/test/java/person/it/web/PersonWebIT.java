package person.it.web;

import static io.restassured.RestAssured.when;
import static io.restassured.RestAssured.with;
import static org.hamcrest.Matchers.emptyOrNullString;
import static org.hamcrest.Matchers.endsWith;
import static org.hamcrest.Matchers.is;

import java.util.HashMap;
import java.util.Map;

import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MediaType;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import com.github.database.rider.core.api.dataset.DataSet;
import com.github.database.rider.junit5.DBUnitExtension;

import io.restassured.RestAssured;

@ExtendWith(DBUnitExtension.class)
public class PersonWebIT {

	private Map<String, String> buildRequestPayload(String firstName, String lastName) {
		Map<String, String> payload = new HashMap<String, String>();
		payload.put("firstName", firstName);
		payload.put("lastName", lastName);
		return payload;
	}

	@BeforeAll
	public static void setup() {
		RestAssured.baseURI = "http://localhost";
		RestAssured.basePath = "/person-web/app/person";
		RestAssured.port = 8080;
	}

	@Test
	public void postCreated() {
		Map<String, String> payload = buildRequestPayload("Nikola", "Tesla");
		String responseBody = with().header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
				.body(payload).when().post().then().statusCode(201).header(HttpHeaders.LOCATION, endsWith("/1"))
				.extract().asString();
		Approvals.verify(responseBody);
	}

	@Test
	@DataSet(value = "yml/nikolaTesla.yml")
	public void putOk() {
		Map<String, String> payload = buildRequestPayload("Nikola-UPDATED", "Tesla-UPDATED");
		String responseBody = with().header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
				.body(payload).when().put("/{id}", 1).then()
				.statusCode(200).extract().asString();
		Approvals.verify(responseBody);
	}

	@Test
	@DataSet(value = "yml/albertEinstein.yml")
	public void getByIdOk() {
		String responseBody = when().get("/{id}", 2).then().statusCode(200).extract().asString();
		Approvals.verify(responseBody);
	}

	@Test
	@DataSet(value = "yml/empty.yml")
	public void getByIdNotFound() {
		when().get("/{id}", 999).then().statusCode(404).body(is(emptyOrNullString()));
	}

	@Test
	@DataSet(value = "yml/carlSagan.yml")
	public void deleteNoContent() {
		when().delete("/{id}", 3).then().statusCode(204).body(is(emptyOrNullString()));
	}

	@Test
	@DataSet(value = "yml/everybody.yml")
	public void getAllOk() {
		String responseBody = when().get().then().statusCode(200).extract().asString();
		Approvals.verify(responseBody);
	}

}
