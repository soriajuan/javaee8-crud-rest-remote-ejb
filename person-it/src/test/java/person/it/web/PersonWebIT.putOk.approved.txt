{"firstName":"Nikola-UPDATED","id":1,"lastName":"Tesla-UPDATED"}